src/fdc/D88DiskImage.cc

```diff
@@ -0,0 +1,350 @@
+}
+
+bool D88DiskImage::readD88Track(uint8_t track, uint8_t side)
+{
+    int32_t track_ = singleSided ? track : track * 2 + side;
+
+    if (track_ >= 160)
+        return false;
+
+	// check readed track.
+	if (cachedTrackNo == track_)
+		return true;
+	cachedTrackNo = -1;
+
+    auto trackOffs = trackOffsets[track_];
+    auto nextTrackOffs = trackOffsets[track_ + 1];
+    if (nextTrackOffs < 1)
+        nextTrackOffs = diskSize;
+    auto trackSize = nextTrackOffs - trackOffs;
+
+    if (trackOffs==0 || trackSize<1)
+        return false;
+
+	if (trackData.size() < trackSize)
+	{
+		trackData.resize(trackSize);
+	}
+	trackSectors.clear();
+
+	/* Read TrackData */
+    file->seek(trackOffs);
+	file->read(std::span{ trackData });
+
+    uint16_t inTrackOffset = 0;
+    while(inTrackOffset < trackSize)
+    {
+		D88SectorHeader* pHeader = (D88SectorHeader*)(trackData.data() + inTrackOffset);
+		D88SectorData sector;
+
+		int nn = pHeader->n & 3;
+		sector.c = pHeader->c;
+		sector.h = pHeader->h;
+		sector.r = pHeader->r;
+		sector.n = pHeader->n;
+		sector.deleteFlag = pHeader->deleteFlag;
+		sector.status = pHeader->status;
+		sector.sectorSize = 128 << nn;
+		sector.dataOffset = inTrackOffset + 16;
+		sector.dataSize = pHeader->sectorDataSize;
+
+		inTrackOffset += (16 + pHeader->sectorDataSize);
+		trackSectors.push_back(sector);
+    }
+
+	cachedTrackNo = track_;
+    return true;
+}
+
+// Disk
+void D88DiskImage::readTrack(uint8_t track, uint8_t side, RawTrack& output)
+{
+	assert(side < 2);
+	if (singleSided && side) {
+		// NoExists Side
+		output.clear(RawTrack::STANDARD_SIZE);
+        return;
+	}
+
+	if (!readD88Track(track, side))
+	{
+		// Track Not Found
+		output.clear(RawTrack::STANDARD_SIZE);
+        return;
+	}
+
+	// # preamble (146)
+	//
+	// GAP4a   0x4E :  80
+	// SYNC    0x00 :  12
+	// IAM     0xC2 :   3
+	//         0xFC :   1
+	// GAP1    0x4E :  50
+	//
+	// # sector (62 + datasize + gap3)
+	//
+	// SYNC    0x00 :  12
+	// IDAM    0xA1 :   3
+	//         0xFE :   1
+	// CHRN      xx :   4
+	// CRC       xx :   2
+	// GAP2    0x4E :  22
+	//
+	// SYNC    0x00 :  12
+	// DAM     0xA1 :   3
+	//         0xFB :   1 (DDAM:0xF8)
+	// data      xx : 128 << N (512)
+	// CRC       xx :   2
+	// GAP3    0x4E :  84 <- 512 byte 9 sec
+	//
+	// # postamble
+	// GAP4b   0x4E : 182 <- 512 byte 9 sec
+	int trackLen = 146;
+	for(const auto& sct: trackSectors)
+	{
+		trackLen += 62 + 84 + sct.sectorSize;
+	}
+	trackLen += 182;
+	output.clear(trackLen);
+
+	int idx = 0;
+	auto write = [&](unsigned n, uint8_t value) {
+		repeat(n, [&] { output.write(idx++, value); });
+	};
+	auto trackDataSpan = std::span{trackData};
+
+	// preamble
+	write(80, 0x4E); // gap4a
+	write(12, 0x00); // sync
+	write( 3, 0xC2); // index mark (1)
+	write( 1, 0xFC); //            (2)
+	write(50, 0x4E); // gap1
+
+	// sectors
+	for(const auto& sct: trackSectors)
+	{
+		if (sct.status == D88_STATUS_NO_ADDRESS_MARK)
+		{
+			continue;
+		}

```

Comment:
Above in the calculation of `trackLen` you assume that all sectors contribute `62 + 84 + sct.sectorSize` bytes. But because of this `continue` that's not correct.
* Maybe we should instead write 0x4E bytes???
* Or maybe this is fine??? Then there will be a larger gap at the end of the track ... though with uninitialized data ... that's not good.

--

There's a similar problem below, with the `continue` on "NO_DATA_MARK".